net: pkt_filter: Add priority filters to enable quality of service

ClaCodes opened the pull request:
## Quality of Service for Ethernet
To enable quality-of-service (QoS) applications, add net_pkt_filters with `result` `NET_CONTINUE`, that allow to change the priority of an incoming packet dynamically. Then the processing of a network packet will happen on the respective traffic-class-queue. This enables a gradual degrading of availabilit of a service based on its priority (user application configurable).

## Results (Sample in this PR)
### With QoS

c (x) := command service for priority x (high means higher priority)

e (x) := echo service for priority x (high means higher priority)

<img width="1564" height="1305" alt="image" src="https://github.com/user-attachments/assets/8aab70dd-4feb-42ee-906f-318906944882" />



### Without QoS
c (x) := command service for priority x (high means higher priority)

e (x) := echo service for priority x (high means higher priority)
<img width="1564" height="1305" alt="image" src="https://github.com/user-attachments/assets/c03e7a1f-9d15-438a-bc48-499be193f7b8" />


## Future Work
- [ ] Implement deferring before `net_conn_input`
- [ ] QoS for UDP
- [ ] QoS for TCP

ClaCodes committed:
net: pkt_filter: Filters to modify priority to incoming packets

To enable quality-of-service (QoS) applications, allow filters to modify
the priority of an incoming packet. This allows that processing can be
done on tc-queue-threads with different priorities.

Signed-off-by: Cla Mattia Galliard <clamattia@gmail.com>

rlubos commented on subsys/net/ip/net_core.c:
```diff
@@ -137,29 +136,47 @@ static inline enum net_verdict process_data(struct net_pkt *pkt)
+			net_queue_rx(net_pkt_iface(pkt), pkt);
+		}
+
+	} while (true);
```
TBH I'm not too fond with this loop approach, it'll add extra processing time for each packet, regardless of whether this TC-requeueing feature is used.

ClaCodes replied:
I did this, because there was precedent (`again` loop).

rlubos replied:
True, but that was only the case with virtual interfaces (will need @jukkar to get more info on how this worked), now every single packet will need to loop after l2 processing at least.

ClaCodes replied:
This is not essential to the RFC. Alternatively could just update priority, conditionally requeue and return `NET_OK` in-place.

rlubos commented on subsys/net/ip/net_core.c:
```diff
@@ -137,29 +136,47 @@ static inline enum net_verdict process_data(struct net_pkt *pkt)
-	default:
-		NET_DBG("Dropping pkt %p", pkt);
+
+		update_priority(pkt);
```
Plus, I wonder, shouldn't there be a spearate `update_priority()` function for each layer, by design? Just thinking out loud, but I imagine that after L3 has processed the packet, we shouldn't be using L2-specific rules anymore?

ClaCodes replied:
Good catch. Will rethink this.

rlubos commented on subsys/net/ip/net_core.c:
```diff
@@ -71,39 +71,39 @@ static inline enum net_verdict process_data(struct net_pkt *pkt)
 	if (!net_pkt_is_raw_processed(pkt)) {
 		net_pkt_set_raw_processed(pkt, true);
 		net_packet_socket_input(pkt, ETH_P_ALL, SOCK_RAW);
+		return NET_CONTINUE;
```
I don't think packet-socket processing would require to recalculate the packet priority?

rlubos reviewed (commented):
Some initial feedback (on the topmost 2 commits).

ClaCodes commented on subsys/net/ip/net_tc.c:
```diff
@@ -65,6 +65,21 @@ static struct net_traffic_class tx_classes[NET_TC_TX_COUNT];
+	int priority;
+	int desired_priority;
+
+	thread_priority = tx_tc2thread(tc);
```
Good catch. Fixed.

ClaCodes commented on subsys/net/ip/net_core.c:
```diff
@@ -137,29 +136,47 @@ static inline enum net_verdict process_data(struct net_pkt *pkt)
+		update_priority(pkt);
+
+		if (!being_processed_by_correct_thread(pkt)) {
+			net_queue_rx(net_pkt_iface(pkt), pkt);
```
Absolutely. Fixed.

ClaCodes commented:
- Rebased

Note: contains all commits from https://github.com/zephyrproject-rtos/zephyr/pull/93246 and adds two commit

rlubos reviewed (commented):
I think packet rescheduling logic turned out pretty straightforward and looks fine, however I think we need to think on how would the priority updating look like in practice.

tpambor commented:
@ClaCodes Could you rebase this PR as the prerequisites https://github.com/zephyrproject-rtos/zephyr/pull/93246 and https://github.com/zephyrproject-rtos/zephyr/pull/93050 have been merged

ClaCodes committed:
samples: net: QoS: Add quality of service sample

Add a sample, that shows the power of quality of service in Zephyr.

Signed-off-by: Cla Mattia Galliard <clamattia@gmail.com>

ClaCodes committed:
net: tc: allow to spread threads by more then 1 priority level

Allow to spread tc threads by more then one priority level and cleanup the
computation of the priority.

Signed-off-by: Cla Mattia Galliard <clamattia@gmail.com>

jukkar commented on subsys/net/ip/net_core.c:
```diff
@@ -90,9 +99,12 @@ static inline enum net_verdict process_data(struct net_pkt *pkt)
 				net_stats_update_processing_error(
 							net_pkt_iface(pkt));
 			}
-
```
same here, and actually it is be preferred that there is an empty line after closing block (so the original code is ok)

jukkar commented on subsys/net/ip/net_core.c:
```diff
@@ -90,9 +99,12 @@ static inline enum net_verdict process_data(struct net_pkt *pkt)
-
 			return ret;
 		}
+		if (!being_processed_by_correct_thread(pkt)) {
```
like here, it would be preferred if there is empty line before the added `if` statement

jukkar commented on subsys/net/ip/net_core.c:
```diff
@@ -129,6 +141,7 @@ static inline enum net_verdict process_data(struct net_pkt *pkt)
 	return NET_DROP;
 }
 
+
```
Please remove this extra empty line

ClaCodes commented on subsys/net/ip/net_core.c:
```diff
@@ -77,7 +87,6 @@ static inline enum net_verdict process_data(struct net_pkt *pkt)
 	if (!pkt->frags) {
 		NET_DBG("Corrupted packet (frags %p)", pkt->frags);
 		net_stats_update_processing_error(net_pkt_iface(pkt));
-
```
done

ClaCodes commented on subsys/net/pkt_filter/base.c:
```diff
@@ -114,6 +114,10 @@ static enum net_verdict evaluate(sys_slist_t *rule_head, struct net_pkt *pkt)
 
 	SYS_SLIST_FOR_EACH_CONTAINER(rule_head, rule, node) {
 		if (apply_tests(rule, pkt) == true) {
+			if (rule->result == NET_CONTINUE) {
```
done: https://github.com/zephyrproject-rtos/zephyr/pull/93341/commits/0f0ed9a70e1a8adc269f7f0bff14ef0ad6b1aa5f

ClaCodes commented on samples/net/QoS/ethernet/src/main.c:
```diff
@@ -0,0 +1,459 @@
+	copy_mac_to(dest);
+
+	pkt = net_pkt_rx_alloc_with_buffer(iface, MTU, AF_UNSPEC, 0, K_NO_WAIT);
+	if (!pkt) {
```
done

ClaCodes commented:
- any suggestion for the doc build?

rlubos commented on include/zephyr/net/net_pkt_filter.h:
```diff
@@ -242,6 +243,14 @@ bool npf_remove_all_rules(struct npf_rule_list *rules);
 		.tests = { FOR_EACH(Z_NPF_TEST_ADDR, (,), __VA_ARGS__) }, \
 	}
 
+#define NPF_PRIORITY(_name, _priority, ...)                                                        \
```
A minimal example of how to create such priority rule could possibly be added to the pkt filter documentation (https://github.com/zephyrproject-rtos/zephyr/blob/main/doc/connectivity/networking/api/net_pkt_filter.rst).

rlubos commented on samples/net/QoS/ethernet/README.rst:
```diff
@@ -0,0 +1,137 @@
+The purpose of this sample is to show quality-of-service on the ethernet layer.
+
+The source code for this sample application can be found at:
+:zephyr_file:`samples/net/QoS`.
```
It's missing `/ethernet/` in the path

ClaCodes replied:
fixed

ClaCodes replied:
Thanks for pointing out. Removed this sentence completely.

rlubos reviewed (commented):
> I tested the sample only on native sim. The sample fakes a network interface and fake-receives packets on it. It then emulates working on it using native-sim specific APIs. It naturally does not build for other platforms (failing tests). Must every sample run on all platforms?

It's fine but should be documented in the readme file if it really can't work with other platforms. Building n in the CI could be limited in `sample.yaml` as already suggested.

ClaCodes committed:
doc: net_pkt_filter: Document priority rules

Add documentation for newly added priority rules.

Signed-off-by: Cla Mattia Galliard <clamattia@gmail.com>

ClaCodes commented:
- rebased
- added docs
- applied suggestions

ClaCodes commented on samples/net/qos/ethernet/src/main.c:
```diff
@@ -0,0 +1,435 @@
+{
+	LOG_INF("sending service %d, pkt %p", net_pkt_ll_proto_type(pkt), pkt);
+
+	posix_cpu_hold(simulated_work_time);
```
No, this would defeat the purpose, as the scheduler would then allow lower priority packets of being processed destoying the demonstration value.

ClaCodes commented on samples/net/QoS/ethernet/README.rst:
```diff
@@ -0,0 +1,137 @@
+   :goals: build
+   :compact:
+
+After the sample starts, it fakes the reception of networkt pakets and prints
```
fixed

ClaCodes commented on doc/connectivity/networking/api/net_pkt_filter.rst:
```diff
@@ -86,6 +93,42 @@ Another (less efficient) way to achieve the same result could be:
     }
 
+This example assigns priorities to different network traffic. It gives network
+control priority (``NET_PRIORITY_CA``) to the ``ptp`` packets, critical
```
Nice catch, fixed.

ClaCodes commented:
- rebased
- adjusted docs as suggested

ClaCodes commented:
- adjusted doc sample to show rule with two match criteria and renamed variabbles for clearity

tpambor reviewed (dismissed):
Nice work. I tested this on hardware with a custom netload test suite and it works as expected.

pdgendt commented on subsys/net/ip/Kconfig:
```diff
@@ -294,6 +294,20 @@ config NET_TC_THREAD_PRIO_CUSTOM
 	  Customise net threads priority by each.
 
 if NET_TC_THREAD_PRIO_CUSTOM
+config NET_TC_TX_THREAD_PRIO_SPREAD
```
Nit, add newline
```suggestion
if NET_TC_THREAD_PRIO_CUSTOM

config NET_TC_TX_THREAD_PRIO_SPREAD
```

ClaCodes committed:
net: shell: filter: Update table

Adjust table to print newly available information.

Signed-off-by: Cla Mattia Galliard <clamattia@gmail.com>

ClaCodes committed:
samples: net: pkt_filter: Add priority filters

Add some priority filters to the sample for demonstration purpose.

Signed-off-by: Cla Mattia Galliard <clamattia@gmail.com>

ClaCodes commented:
- rebased
- fixed nits and typos
- Added to `net filter` shell command -> Found an issue here. Next PR or will create proper issue
- Added priority filters to the existing sample in `net/pkt_filter`. Running `net filter` on that sample now gives:
```
Rule  Type        Verdict   Pkt-Prio  Queue  Thread-Prio  Tests
[ 1]  recv        CONTINUE         1      0            1  1    iface[1]
[ 2]  recv        CONTINUE         7      1         SKIP  2    iface[1],eth type[0x88f7]
[ 3]  recv        CONTINUE         2      0            1  2    iface[1],eth type[0x8100]
[ 4]  recv        CONTINUE         1      0            1  2    iface[2],eth vlan type[0x0806]
[ 5]  recv        CONTINUE         1      0            1  2    iface[3],eth vlan type[0x0806]
[ 6]  recv        OK             N/A    N/A          N/A  3    iface[2],eth vlan type[0x0800],size max[200]
[ 7]  recv        OK             N/A    N/A          N/A  3    iface[3],eth vlan type[0x0800],size min[100]
[ 8]  recv        OK             N/A    N/A          N/A  1    iface[1]
[ 9]  recv        OK             N/A    N/A          N/A  2    iface[2],eth vlan type[0x0806]
[10]  recv        OK             N/A    N/A          N/A  2    iface[3],eth vlan type[0x0806]
[11]  recv        DROP           N/A    N/A          N/A  0    
```

ClaCodes commented on samples/net/qos/ethernet/src/main.c:
```diff
@@ -0,0 +1,428 @@
+static NPF_PRIORITY(priority_rule_129, NET_PRIORITY_IC, npf_129);
+static NPF_PRIORITY(priority_rule_130, NET_PRIORITY_NC, npf_130);
+
+static uint32_t simulated_work_time; /* data race ! */
```
Yes, as I noted with the comment. But it is very unlikely, that is written while readers are active, and I would like to avoid them blocking each other.

pdgendt reviewed (approved):
In the future, try to avoid force pushing a rebase combined with changes. This makes the "compare" function in Github useless.
